mobile_verifier/src/owner_shares.rs

```diff
@@ -3,17 +3,106 @@ use crate::{
     speedtests::{Average, SpeedtestAverages},
 };
 use chrono::{DateTime, Duration, TimeZone, Utc};
-use file_store::traits::TimestampEncode;
+use file_store::{mobile_transfer::ValidDataTransferSession, traits::TimestampEncode};
+use futures::{Stream, StreamExt};
 use helium_crypto::PublicKeyBinary;
 use helium_proto::services::{
     follower::{self, follower_gateway_resp_v1::Result as GatewayResult, FollowerGatewayReqV1},
     poc_mobile as proto, Channel,
 };
 use lazy_static::lazy_static;
 use rust_decimal::prelude::*;
+use rust_decimal_macros::dec;
 use std::collections::HashMap;
 use std::ops::Range;
 
+pub struct TransferRewards {
+    scale: Decimal,
+    rewards: HashMap<PublicKeyBinary, Decimal>,
+    pub remaining_rewards: Decimal,
+}
+
+impl TransferRewards {
+    pub fn empty(epoch: &Range<DateTime<Utc>>) -> Self {
+        Self {
+            scale: Decimal::ONE,
+            rewards: HashMap::new(),
+            remaining_rewards: get_scheduled_tokens(epoch.start, epoch.end - epoch.start).unwrap(),
+        }
+    }
+
+    pub fn reward(&self, hotspot: &PublicKeyBinary) -> Decimal {
+        self.rewards.get(hotspot).copied().unwrap_or(Decimal::ZERO) * self.scale
+    }
+
+    pub async fn from_transfer_sessions(
+        mobile_price: Decimal,
+        transfer_sessions: impl Stream<Item = ValidDataTransferSession>,
+        epoch: &Range<DateTime<Utc>>,
+    ) -> Result<Self, ClockError> {
+        tokio::pin!(transfer_sessions);
+
+        let mut data_transfer_reward_sum = Decimal::ZERO;
+        let rewards = transfer_sessions
+            // Accumulate bytes per hotspot
+            .fold(
+                HashMap::<PublicKeyBinary, Decimal>::new(),
+                |mut entries, session| async move {
+                    *entries.entry(session.pub_key).or_default() +=
+                        Decimal::from(bytes_to_dc(session.download_bytes + session.upload_bytes));
+                    entries
+                },
+            )
+            .await
+            .into_iter()
+            // Calculate rewards per hotspot
+            .map(|(pub_key, dc_amount)| {
+                let mobiles = dc_to_mobile_bones(dc_amount, mobile_price);
+                data_transfer_reward_sum += mobiles;
+                (pub_key, mobiles)
+            })
+            .collect();
+
+        let total_rewards = get_scheduled_tokens(epoch.start, epoch.end - epoch.start)?;
+
+        let (scale, remaining_rewards) =
+            if data_transfer_reward_sum / total_rewards > *MAX_DATA_TRANSFER_REWARDS_PERCENT {
+                let scale = *MIN_POC_REWARDS_PERCENT * total_rewards / data_transfer_reward_sum;
```

Comment:
why use `MIN_POC_REWARDS_PERCENT` to determine scale, shouldnt it be using `MAX_DATA_TRANSFER_REWARDS_PERCENT` ?
Example
total_rewards: 100000
total data rewards: 40000 ( 2x the cap of 0.2 )
total poc rewards: 80000

As the data rewards exceeds the 0.2 cap, scale should be 0.5

if we use the current formula:

MIN_POC_REWARDS_PERCENT * total_rewards / total data rewards = 2

whereas if we use 

MAX_DATA_TRANSFER_REWARDS_PERCENT * total_rewards / total data rewards = 0.5

Reply:
Yup, not sure how I got that math wrong :man_facepalming: